sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/simple/InMemoryBodyCollector.java

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.netty.implementation.simple;
+
+import com.azure.core.util.BinaryData;
+import io.netty.buffer.ByteBuf;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+public class InMemoryBodyCollector implements SimpleBodyCollector {
+
+    private volatile ByteArrayOutputStream outputStream;
+
+    @Override
+    public void collect(ByteBuf buffer, boolean isLast) {
+        if (buffer.isReadable()) {
+            ensureStream();
+            try {
+                buffer.readBytes(outputStream, buffer.readableBytes());
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    private void ensureStream() {
+        if (outputStream == null) {
+            synchronized (this) {
+                if (outputStream == null) {
+                    outputStream = new ByteArrayOutputStream();
+                }
+            }
+        }
```

Comment:
A recommendation in "Effective Java (EJ)" while following DCL is to introduce a local variable. As per the EJ, such a version may improve the perf by about 25% compared to the version without the local variable. 

The code would be (pattern copied from EJ)

```java
public void collect(ByteBuf buffer, boolean isLast) {
    if (buffer.isReadable()) {
        ByteArrayOutputStream result = ensureStream();
        try {
            buffer.readBytes(result, buffer.readableBytes());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}

private ByteArrayOutputStream ensureStream() {
    ByteArrayOutputStream result = outputStream;
    if (result == null) {
        synchronized (this) {
            result = outputStream;
            if (result == null) {
                outputStream = result = new ByteArrayOutputStream();
            }
        }
    }
    return result;
}

```